ratatui-core/src/buffer/buffer.rs

```diff
@@ -345,19 +345,25 @@ impl Buffer {
             .map(|symbol| (symbol, symbol.width() as u16))
             .filter(|(_symbol, width)| *width > 0)
             .map_while(|(symbol, width)| {
-                remaining_width = remaining_width.checked_sub(width)?;
+                if width > remaining_width {
+                    return None;
+                }
+                remaining_width = remaining_width.saturating_sub(width);
                 Some((symbol, width))
             });
+
         let style = style.into();
         for (symbol, width) in graphemes {
             self[(x, y)].set_symbol(symbol).set_style(style);
-            let next_symbol = x + width;
-            x += 1;
-            // Reset following cells if multi-width (they would be hidden by the grapheme),
-            while x < next_symbol {
-                self[(x, y)].reset();
-                x += 1;
+            // For each grapheme, advance x by exactly its width
+            // This ensures that multi-width characters like Chinese characters
+            // are properly handled without adding extra whitespace
+            for i in 1..width {
+                if x + i < self.area.right() {
+                    self[(x + i, y)].reset();
+                }
             }
+            x += width;
```

Comment:
This is definitely better, but it's not clear that the if check is required (the purpose of the map_while above is to exclude characters which would make `x + width >= area.right()`